### ISSUE TITLE
Question: Let cabal handle cbits compilation?

### ISSUE THREAD

**Issue by lancelet** [google-research/dex-lang#123]
Q: Is there a reason to compile `libdex.c` as a separate step in the `makefile`?

I'm keen to try letting Cabal compile `libdex.c`, in the pattern that is often used elsewhere (eg. see this example in [the bytestring package](https://github.com/haskell/bytestring/blob/e6cb01e2ec0bfdd19298418c85f220925a9fa307/bytestring.cabal#L131-L132)).

However, since it might well be set up the way it is for good reasons, I thought I'd check first before I give it a go. Is this a good idea? :-)

btw: my motivation is to allow `stack install` to work properly. Currently, if I `stack install` and then try to run `dex` from a different directory, I see this:
```
$ dex
dyld: Library not loaded: cbits/libdex.so
  Referenced from: /Users/jsm/.local/bin/dex
  Reason: image not found
[2]    48521 abort      dex
```
Since it's trying to load `libdex.so` dynamically rather than statically linking that code.

**Comment by dougalm** [comment]
> Q: Is there a reason to compile libdex.c as a separate step in the makefile?

No reason at all. I forget if I tried to make Cabal compile it and gave up or if I didn't even know it was possible.

> my motivation is to allow stack install to work properly

That would be amazing!

While you're thinking about linking, maybe you can find a better alternative to [this hack](https://github.com/google-research/dex-lang/blob/main/src/lib/LLVMExec.hs#L32-L33). The problem is that we need the code generated by LLVM to be able to call into libdex, but of course Cabal doesn't know about it.

**Comment by juhp** [comment]
Dex looks very interesting.

```
$ git rev-parse --short HEAD
79125a7
$ stack install
: 
:
[27 of 27] Compiling LiveOutput
Preprocessing executable 'dex' for dex-0.1.0.0..
Building executable 'dex' for dex-0.1.0.0..
Warning: the following files would be used as linker inputs, but linking is not being done: cbits/libdex.so
[1 of 1] Compiling Main
Linking .stack-work/dist/x86_64-linux-tinfo6/Cabal-2.4.0.1/build/dex/dex ...
gcc: error: cbits/libdex.so: No such file or directory
`gcc' failed in phase `Linker'. (Exit code: 1)
```

This is on Fedora 32 fwiw, I got similar result with `cabal new-build`.

**Comment by lancelet** [comment]
@juhp I saw that error too (and it's quite relevant to this issue!)... it's because the C components are being compiled as a separate step before stack or cabal are run.

Have you tried running `make` instead? Currently, the C code has to be compiled first, which is handled by the `makefile`.

**Comment by juhp** [comment]
Thanks @lancelet 
I tried building libdex with make first, but still get the same error afterwards with stack or cabal.

**Comment by lancelet** [comment]
I'll have to investigate what Cabal is doing a bit more thoroughly to see if my initial suggestion is possible. I hadn't appreciated that the C code is being called by jitted LLVM, and is more part of the "dex runtime" (if I can call it that) than the dex compiler.

I can get Cabal to build the C code, but LLVM is relatively new territory for me. I'm not sure if it can be pointed at code which has been statically compiled into `dex`. I know Stephen Diehl has [written quite a lot about JIT compilation with LLVM](https://www.stephendiehl.com/llvm/), so I'll see if his material contains any hints about doing this.

I also discovered that `prelude.dx` has to be available if the `dex` binary is run from somewhere outside the source tree (well, duh... 😄). Not a difficult problem though.

@juhp after you have run `make`, does it produce the `cbits/libdex.so` file? eg (on macOS):
```
$ make
... lots of compilation ...
$ ls -l cbits/libdex.so
-rwxr-xr-x  1 jsm  staff  13328 29 Jun 20:09 cbits/libdex.so
$ stack exec dex -- --help
Usage: dex COMMAND [--backend ARG] [--prelude FILE] [--logto FILE]

Available options:
  --backend ARG            Backend (LLVM|JAX|interp)
  --prelude FILE           Prelude file (default: "prelude.dx")
  --logto FILE             File to log to
  -h,--help                Show this help text

Available commands:
  repl
  web
  watch
  script
```
On my end, `make` also runs `stack` (via the `all` target in the `makefile`). However, if I make a change, I can run `stack` by itself afterwards, provided `cbits/libdex.so` already exists.

**Comment by juhp** [comment]
Maybe it works on MacOS but not Linux?

**Comment by chessai** [comment]
I have a PR which fixes this

**Comment by juhp** [comment]
Thanks that's good to know - I suppose as such your .cabal changes are independent of nix.

**Comment by apaszke** [comment]
I've actually implemented this in #193.

**Comment by chessai** [comment]
Nice, thanks. I had forgotten about this. I will PR nix stuff separately.
